### PR TITLE
Enhance debug logs

### PR DESCRIPTION
**Pull request by lowercase00**
There were reports where a worker wouldn't fetch new jobs.
Since one possible cause is the infinite hanging of the `BLPOP` command,
this adds a couple of logs surrounding the `dequeue_any` function,
that interacts with Redis using `BLPOP`,

### CODE REVIEW

File: rq/worker.py [hunk 1]
```diff
@@ -690,10 +690,12 @@ def dequeue_job_and_maintain_ttl(self, timeout):
                 if self.should_run_maintenance_tasks:
                     self.run_maintenance_tasks()
 
+                self.log.debug(f"Dequeing jobs on queues {self._ordered_queues} and timeout {timeout}")
```
Comment (selwin) [inline rq/worker.py:693]: Sorry, minor nitpick `Dequeueing job from self._ordered_queues with timeout of {}`

File: rq/worker.py [hunk 2]
```diff
@@ -690,10 +690,12 @@ def dequeue_job_and_maintain_ttl(self, timeout):
...
                 result = self.queue_class.dequeue_any(self._ordered_queues, timeout,
                                                       connection=self.connection,
                                                       job_class=self.job_class,
                                                       serializer=self.serializer)
+                self.log.debug(f"Dequeued job ID {result[1]} on queue {result[0]}")
```
Comment (selwin) [inline rq/worker.py:698]: `Dequeued job {result[1]} from {result[0]}`

File: rq/worker.py [hunk 3]
```diff
@@ -1098,17 +1105,23 @@ def perform_job(self, job: 'Job', queue: 'Queue'):
         """Performs the actual work of a job.  Will/should only be called
         inside the work horse's process.
         """
+        self.log.debug("Starting process to perform a job...")
         push_connection(self.connection)
+        self.log.debug("Connection pushed.")
```
Comment (selwin) [inline rq/worker.py:1110]: I think we can get rid of these two statements.

File: rq/worker.py [hunk 4]
```diff
@@ -1084,12 +1089,14 @@ def handle_job_success(self, job: 'Job', queue: 'Queue', started_job_registry: S
 
     def execute_success_callback(self, job: 'Job', result):
         """Executes success_callback with timeout"""
+        self.log.debug("Running Success Callbacks.")
```
Comment (selwin) [inline rq/worker.py:1092]: `Running success callbacks for {job.id}`

File: rq/worker.py [hunk 5]
```diff
@@ -1084,12 +1089,14 @@ def handle_job_success(self, job: 'Job', queue: 'Queue', started_job_registry: S
...
         job.heartbeat(utcnow(), CALLBACK_TIMEOUT)
         with self.death_penalty_class(CALLBACK_TIMEOUT, JobTimeoutException, job_id=job.id):
             job.success_callback(job, self.connection, result)
 
     def execute_failure_callback(self, job):
         """Executes failure_callback with timeout"""
+        self.log.debug("Running Failure Callbacks.")
```
Comment (selwin) [inline rq/worker.py:1099]: `Running Failure Callbacks for {job.id}`

File: rq/worker.py [hunk 6]
```diff
@@ -1098,17 +1105,23 @@ def perform_job(self, job: 'Job', queue: 'Queue'):
...
         """
+        self.log.debug("Starting process to perform a job...")
         push_connection(self.connection)
+        self.log.debug("Connection pushed.")
 
         started_job_registry = queue.started_job_registry
+        self.log.debug("Started Job Registry set.")
 
         try:
+            self.log.debug("Preparing for execution...")
```
Comment (selwin) [inline rq/worker.py:1116]: I think we can remove this statement, there's already a logging statement in `prepare_job_execution`

File: rq/worker.py [hunk 7]
```diff
@@ -1123,6 +1136,7 @@ def perform_job(self, job: 'Job', queue: 'Queue'):
                                     queue=queue,
                                     started_job_registry=started_job_registry)
         except:  # NOQA
+            self.log.debug("Job Execution failed.")
```
Comment (selwin) [inline rq/worker.py:1139]: `Job <ID> raised an exception`

File: rq/worker.py [hunk 8]
```diff
@@ -1140,6 +1154,7 @@ def perform_job(self, job: 'Job', queue: 'Queue'):
 
             self.handle_job_failure(job=job, exc_string=exc_string, queue=queue,
                                     started_job_registry=started_job_registry)
+            self.log.debug("Handling Exception.")
```
Comment (selwin) [inline rq/worker.py:1157]: Can we move this statement into `self.handle_exception`?

**Comment by lowercase00** [comment]
@selwin I think all comments are resolved now. I'm thinking that we should also add some logs to the `queue` to help capture the application's state when pushing the job to a queue, especially at `push_job_id` that way we could check whether the `rpush` has any issues. Thoughts?

**Comment by lowercase00** [comment]
@selwin take a look at the suggestion I made on https://github.com/rq/rq/pull/1752/commits/55dee09dd7fe4958b01a9ae446f385de38f97862. There are a log of logs, but they should help understanding where the issue may be happening.

File: rq/queue.py [hunk 9]
```diff
@@ -291,11 +305,13 @@ def compact(self):
     def push_job_id(self, job_id: str, pipeline: t.Optional['Pipeline'] = None, at_front=False):
         """Pushes a job ID on the corresponding Redis queue.
         'at_front' allows you to push the job onto the front instead of the back of the queue"""
+        self.log.debug(f"Pushing job {blue(job_id)} into the queue {green(self.name)}, current job count {self.count}")
```
Comment (selwin) [inline rq/queue.py:308]: `self.count` incurs a Redis operation so I'd rather not have job count displayed here.

File: rq/queue.py [hunk 10]
```diff
@@ -291,11 +305,13 @@ def compact(self):
...
         connection = pipeline if pipeline is not None else self.connection
         if at_front:
-            connection.lpush(self.key, job_id)
+            result = connection.lpush(self.key, job_id)
         else:
-            connection.rpush(self.key, job_id)
+            result = connection.rpush(self.key, job_id)
+        self.log.debug(f"Pushed job {blue(job_id)} into the queue {green(self.name)}, total job count after pushing job: {result}")
```
Comment (selwin) [inline rq/queue.py:314]: `Pushed job {blue(job_id)} into {green(self.name)}, {result} job(s) are in in queue.`

File: rq/queue.py (hunk 9 above)
Comment (lowercase00) [reply rq/queue.py:308]: We could have a toggle for "debug mode" to avoid the extra request while not debugging, what do you think? Without the count we would lose the ability to check whether the list was actually increased as it should.

File: rq/worker.py [hunk 11]
```diff
@@ -1089,14 +1089,14 @@ def handle_job_success(self, job: 'Job', queue: 'Queue', started_job_registry: S
...
-        self.log.debug("Running Success Callbacks.")
+        self.log.debug(f"Running success callbacks for {job.id}")
         job.heartbeat(utcnow(), CALLBACK_TIMEOUT)
         with self.death_penalty_class(CALLBACK_TIMEOUT, JobTimeoutException, job_id=job.id):
             job.success_callback(job, self.connection, result)
 
     def execute_failure_callback(self, job):
         """Executes failure_callback with timeout"""
-        self.log.debug("Running Failure Callbacks.")
+        self.log.debug(f"Running Failure Callbacks for {job.id}")
```
Comment (selwin) [inline rq/worker.py:1099]: Nitpick: `failure callbacks` should not have uppercase as first letters :)

File: rq/queue.py [hunk 12]
```diff
@@ -699,10 +715,13 @@ def lpop(cls, queue_keys, timeout, connection: t.Optional['Redis'] = None):
         if timeout is not None:  # blocking variant
             if timeout == 0:
                 raise ValueError('RQ does not support indefinite timeouts. Please pick a timeout value > 0')
+            logger.debug(f"Starting blocking pop operation for queues {green(queue_keys)} with timeout of {timeout}")
```
Comment (selwin) [inline rq/queue.py:718]: Nitpick: we should make the blocking pop terminology consistent. `Starting BLPOP for queues {green(queue_keys)} with timeout of {timeout}`

File: rq/worker.py (hunk 11 above)
Comment (lowercase00) [reply rq/worker.py:1099]: "Should" is a strong word to use :) But got your point will change that.

File: rq/queue.py (hunk 9 above)
Comment (lowercase00) [reply rq/queue.py:308]: Introduced a `debug` param to the queue's `__init__`, and then logged the count only if the debug is `True` (defaults to `False`). This will avoid extra operations when not debugging while maintaining full visibility over the most relevant operations.

File: rq/queue.py [hunk 13]
```diff
@@ -305,13 +306,15 @@ def compact(self):
     def push_job_id(self, job_id: str, pipeline: t.Optional['Pipeline'] = None, at_front=False):
         """Pushes a job ID on the corresponding Redis queue.
         'at_front' allows you to push the job onto the front instead of the back of the queue"""
-        self.log.debug(f"Pushing job {blue(job_id)} into the queue {green(self.name)}, current job count {self.count}")
+        self.log.debug(f"Pushing job {blue(job_id)} into the queue {green(self.name)}.")
+        if self.debug_mode:
```
Comment (selwin) [inline rq/queue.py:310]: Sorry, do you mind deleting this? I don't think we need this as we'll know the job count after job is pushed, sorry :(

File: rq/queue.py (hunk 9 above)
Comment (lowercase00) [reply rq/queue.py]: Undone.

File: rq/queue.py [hunk 14]
```diff
@@ -699,10 +714,13 @@ def lpop(cls, queue_keys, timeout, connection: t.Optional['Redis'] = None):
         if timeout is not None:  # blocking variant
             if timeout == 0:
                 raise ValueError('RQ does not support indefinite timeouts. Please pick a timeout value > 0')
+            logger.debug(f"Starting BLPOP operation for queues {green(queue_keys)} with timeout of {timeout}")
             result = connection.blpop(queue_keys, timeout)
             if result is None:
+                logger.debug(f"BLPOP Timeout, no jobs found on queues {green(queue_keys)}")
                 raise DequeueTimeout(timeout, queue_keys)
             queue_key, job_id = result
+            logger.debug(f"Found job {blue(job_id)} on queue {green(queue_key)}")
```
Comment (selwin) [inline rq/queue.py:723]: One more minor nitpick before I merge this in. Can we slightly rephrase this to `Dequeued job {blue(job_id)} from queue {green(queue_key)}`? This should make it clear that the job has been popped of the queue.

File: rq/worker.py [hunk 15]
```diff
@@ -946,17 +948,20 @@ def prepare_job_execution(self, job: 'Job'):
...
+        self.log.debug(f"Preparing for execution of Job ID {job.id}")
         with self.connection.pipeline() as pipeline:
             self.set_current_job_id(job.id, pipeline=pipeline)
             self.set_current_job_working_time(0, pipeline=pipeline)
+            self.log.debug(f"Set Current Job ID & Working Time")
 
             heartbeat_ttl = self.get_heartbeat_ttl(job)
             self.heartbeat(heartbeat_ttl, pipeline=pipeline)
             job.heartbeat(utcnow(), heartbeat_ttl, pipeline=pipeline)
+            self.log.debug(f"Got Heartbeat. Current TTL is {heartbeat_ttl}")
```
Comment (selwin) [inline rq/worker.py:960]: `Heartbeat set with TTL of {heartbeat_ttl}`

File: rq/worker.py [hunk 16]
```diff
@@ -946,17 +948,20 @@ def prepare_job_execution(self, job: 'Job'):
...
 
             job.prepare_for_execution(self.name, pipeline=pipeline)
             pipeline.execute()
+            self.log.debug(f"Job preparation finished.")
```
Comment (selwin) [inline rq/worker.py:964]: Actually I just realized that the commands for these three log statements are executed in a single `pipeline.execute()` call, meaning when the log is output, the current job ID, working time and heartbeat have not been set yet.

This means we should only log after a `pipeline.execute()`

Comment (lowercase00) [reply rq/worker.py:964]: Not sure I understand how you'd like this to be. My initial idea was just to set logs to understand what was the last active point before a potential issue - those particular logs don't necessarily depend on the pipeline execution. Anyways, fixed the logging text. This is the only one missing, let me know how you'd like this to be.

Comment (selwin) [reply rq/worker.py]: Sorry, then let's just delete these two lines:
1. `self.log.debug(f"Set Current Job ID & Working Time")`
2. `self.log.debug(f"Heartbeat set with TTL of {heartbeat_ttl}")`

I'll merge this PR in after these two are removed.

**Comment by selwin** [comment]
Thanks!
